File: algorithms/0059.spiral_matrix_ii/spiral_matrix_ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(const int n) {
        const int max_num = n * n;
        const int directions[4][4] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}; 
        int row = 0, col = 0, dir_index = 0;
        int cur_num = 1;
        vector<vector<int>> matrix(n, vector<int>(n));
        while (cur_num <= max_num) {
            matrix[row][col] = cur_num;
            ++cur_num;
            int nextRow = row + directions[dir_index][0], nextCol = col + directions[dir_index][1];
            if (nextRow < 0 || nextRow >= n || nextCol < 0 || nextCol >= n || matrix[nextRow][nextCol] != 0) {
                dir_index = (dir_index + 1) % 4;
            }
            row = row + directions[dir_index][0];
            col = col + directions[dir_index][1];
        }
        return move(matrix);
    }
};

File: algorithms/0059.spiral_matrix_ii/spiral_matrix_ii.rs
impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let directions = [(0i32, 1i32), (1, 0), (0, -1), (-1, 0)];
        let mut index = 0usize;
        let mut cnt = 1;
        let (mut row, mut col) = (0i32, 0i32);
        let mut matrix = vec![vec![0; n as usize]; n as usize];

        let mut direction = &directions[index];
        let max = n * n;
        while cnt <= max {
            matrix[row as usize][col as usize] = cnt;
            let (next_row, next_col) = (row + direction.0, col + direction.1);
            cnt += 1;
            if next_row < 0 || next_row >= n || next_col < 0 || next_col >= n || matrix[next_row as usize][next_col as usize] != 0 {
                index += 1;
                direction = &directions[index % 4];
            }
            row += direction.0;
            col += direction.1;
        }
        matrix
    }
}